### PR TITLE
Improve error message for input data to fit.

### LINKED ISSUE #16202: Better error message for Unexpected result of `train_function` (Empty logs). Please use `Model.compile(..., run_eagerly=True)`, or `tf.config.run_functions_eagerly(True)` for more information of where went wrong, or file a issue/bug to `tf.keras`
Hi!

This error message pops up when empty arrays are passed to `Model.fit` or `Model.predict`. The solution suggested in the error message does not work and in some sense is not logical. Either the error message should be updated or the mentioned functions should have additional assertions to check for empty input arrays.

### PR DESCRIPTION
**Pull request by tomrtk**
The open PR for issue #16202 looks to be dead, and having conflicts. So I have added some checks in `DataHandler` to catch empty input and steps. I did not edit the error message in `fit` like the last PR did, as it looks to have been updated already.

This PR improve error message to give more useful feedback:
- if passing empty dataset or array as input data
- if passing `steps_per_epoch=0`

to `Model.fit()` by adding checks in `DataHandler`.

Resolve #16202

### CODE REVIEW

File: keras/engine/data_adapter.py [hunk 1]
```diff
@@ -1271,6 +1271,12 @@ def __init__(
         self._insufficient_data = False
         self._model = model
 
+        if steps_per_epoch == 0:
+            raise ValueError(
+                "Got argument `steps_per_epoch=0` passed to `fit()`."
```
Comment (qlzh727) [inline keras/engine/data_adapter.py:1276]: Let's update this error message to be

"Unexpected value for `steps_per_epoch`. Received value is 0. Please check the docstring for `model.fit()` for supported values."

File: keras/engine/data_adapter.py [hunk 2]
```diff
@@ -1308,6 +1314,9 @@ def __init__(
             strategy, x, steps_per_epoch, class_weight, distribute
         )
 
+        if self._inferred_steps == 0:
+            raise ValueError("Expected input data to `fit()` to be non-empty.")
```
Comment (qlzh727) [inline keras/engine/data_adapter.py:1318]: Note that this method can be called for model.eval/predict as well, so let's remove the model.fit() from the message.

**Review by qlzh727 — CHANGES_REQUESTED** [review]

**Comment by gbaned** [comment]
> there has not been any progress on this pr for a while, so I just want to make sure there is no blockers from my side?

Hi @tomrtk Sorry for the delay, this PR is in awaiting review status. No blockers from your end. Thank you!
Hi @qlzh727 Can you please review this PR ? Thank you!

**Comment by qlzh727** [comment]
Somehow it was blocked in the gerrit review, let me try to kick it again.

**Pull request merged**